mediatek: Add support for GL.iNet X3000 (Spitz AX) and XE3000 (Puli AX)

JeThWifirst opened the pull request:
The GL.iNet X3000 and XE3000 are Wi-Fi 6 5G cellular routers, based on MediaTek MT7981A SoC. The XE3000 is the same device as the X3000, except for an additional battery.

Specifications:
 - SoC: Filogic 820 MT7981A (1.3GHz)
 - RAM: DDR4 512M
 - Flash: eMMC 8G, MicroSD card slot
 - WiFi: 2.4GHz and 5GHz with 6 antennas
 - Ethernet:
   - 1x LAN (10/100/1000M)
   - 1x WAN (10/100/1000/2500M)
 - 5G: Quectel RM520N-GL with two nano-SIM card slots
 - USB: 1x USB 2.0 port
 - UART:
   - 3.3V, TX, RX, GND / 115200 8N1

MAC addresses as verified by OEM firmware:
vendor   OpenWrt    address     source
WAN      eth0       label       factory 0x0a (label)
LAN      eth1       label + 1
2g       phy0-ap0   label + 2   factory 0x04
5g       phy1-ap0   label + 3

Installation via U-Boot rescue:
1. Press and hold reset button while booting the device
2. Wait for the Internet led to blink 5 times
3. Release reset button
4. The rescue page is accessible via http://192.168.1.1
5. Select the OpenWrt sysupgrade image and start upgrade
6. Wait for the router to flash new firmware and reboot

Revert to stock firmware:
1. Download the stock firmware from GL.iNet website
2. Use the method explained above to flash the stock firmware

Switch the modem network port to the USB interface:
1. Connect to the AT commands (/dev/ttyUSB2) port using e.g. minicom: minicom -D /dev/ttyUSB2
2. Check the modem mode with 'AT+QCFG="data_interface"':
  - 0,0 indicates that the network port uses the PCIe interface
  - 1,0 indicates that the network port uses the USB interface
3. Switch to the USB interface if necessary with 'AT+QCFG="data_interface",0,0', and reboot

JeThWifirst committed:
mediatek: Add support for GL.iNet X3000 (Spitz AX) and XE3000 (Puli AX)

The GL.iNet X3000 and XE3000 are Wi-Fi 6 5G cellular routers, based on
MediaTek MT7981A SoC. The XE3000 is the same device as the X3000,
except for an additional battery.

Specifications:
 - SoC: Filogic 820 MT7981A (1.3GHz)
 - RAM: DDR4 512M
 - Flash: eMMC 8G, MicroSD card slot
 - WiFi: 2.4GHz and 5GHz with 6 antennas
 - Ethernet:
   - 1x LAN (10/100/1000M)
   - 1x WAN (10/100/1000/2500M)
 - 5G: Quectel RM520N-GL with two nano-SIM card slots
 - USB: 1x USB 2.0 port
 - UART:
   - 3.3V, TX, RX, GND / 115200 8N1

MAC addresses as verified by OEM firmware:
vendor   OpenWrt    address     source
WAN      eth0       label       factory 0x0a (label)
LAN      eth1       label + 1
2g       phy0-ap0   label + 2   factory 0x04
5g       phy1-ap0   label + 3

Installation via U-Boot rescue:
1. Press and hold reset button while booting the device
2. Wait for the Internet led to blink 5 times
3. Release reset button
4. The rescue page is accessible via http://192.168.1.1
5. Select the OpenWrt sysupgrade image and start upgrade
6. Wait for the router to flash new firmware and reboot

Revert to stock firmware:
1. Download the stock firmware from GL.iNet website
2. Use the method explained above to flash the stock firmware

Switch the modem network port between PCIe and USB interfaces:
1. Connect to the AT commands (/dev/ttyUSB2) port using
e.g. minicom: minicom -D /dev/ttyUSB2
2. Check the current modem mode with 'AT+QCFG="data_interface"':
  - 0,0 indicates that the network port uses the USB interface
  - 1,0 indicates that the network port uses the PCIe interface
3. Switch the active interface with:
  - 'AT+QCFG="data_interface",0,0' to use the USB interface
  - 'AT+QCFG="data_interface",1,0' to use the PCIe interface
4. Reboot

Signed-off-by: Jean Thomas <jean.thomas@wifirst.fr>

mrkiko commented:
Hi!
I was curious about the pinctrl changes. Why where them needed? Thanks a lot for your work.

mrkiko commented:
Also - I may have missed it, but I can't find the call to emmc_copy_config for restoring config after sysupgrade.

dangowrt commented on target/linux/mediatek/patches-6.1/250-pinctrl-mediatek-mt7981-add-additional-mmc-and-uart-groups.patch:
```diff
@@ -0,0 +1,64 @@
+ static int mt7981_emmc_45_pins[] = { 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
+ static int mt7981_emmc_45_funcs[] = { 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, };
+
++static int mt7981_emmc_pins[] = { 0, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
```
Please add `0` to emmc_45_pins above instead, I think this is a bug rather than actually a possible configuration (ie. without pin 0 set to function 2 already by the bootloader this never worked)

JeThWifirst replied:
Indeed, I added `0` to `emmc_45_pins`, and it works. Thanks for the tip.

JeThWifirst commented:
> Hi! I was curious about the pinctrl changes. Why where them needed? Thanks a lot for your work.

The UART pinctrl change is necessary as the UART1 has no RTS/CTS. And the other one was not necessary, it is now removed.

> Also - I may have missed it, but I can't find the call to emmc_copy_config for restoring config after sysupgrade.

Thanks, good catch, this is now added.

aiamadeus commented on target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi:
```diff
@@ -0,0 +1,257 @@
+		#size-cells = <0>;
+
+		phy5: phy@5 {
+			compatible = "ethernet-phy-id67c9.de0a";
```
Please use Clause-45 MDIO instead of hard-coded phy ID.
```
&mdio_bus {
	reset-gpios = <&pio 14 GPIO_ACTIVE_LOW>;
	reset-delay-us = <600>;
	reset-post-delay-us = <20000>;

	phy5: ethernet-phy@5 {
		reg = <5>;
		compatible = "ethernet-phy-ieee802.3-c45";
	};
};
```

JeThWifirst replied:
The `mdio_bus` now uses Clause-45

aiamadeus commented on target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi:
```diff
@@ -0,0 +1,257 @@
+&spi1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spic_pins>;
+	status = "disabled";
```
Can the spi1 node be removed?

JeThWifirst replied:
Indeed, `spi1` and the associated `spic_pins` are removed

aiamadeus commented on target/linux/mediatek/patches-6.1/250-pinctrl-mediatek-mt7981-add-additional-mmc-and-uart-groups.patch:
```diff
@@ -0,0 +1,56 @@
+---
```
Please add patch description.

JeThWifirst committed:
mediatek: mt7981: pinctrl: add additional uart group

Add uart1_3 (pins 26, 27) group to the pinctrl driver for the
MediaTek MT7981 SoC.

Signed-off-by: Jean Thomas <jean.thomas@wifirst.fr>

JeThWifirst committed:
mediatek: mt7981: pinctrl: add additional emmc groups

Add new emmc groups in the pinctrl driver for the
MediaTek MT7981 SoC:
* emmc reset, with pin 15.
* emmc 4-bit bus-width, with pins 16 to 19, and 24 to 25.
* emmc 8-bit bus-width, with pins 16 to 25.

The existing emmc_45 group is kept for legacy reasons, even
if this is the union of emmc_reset and emmc_8 groups.

Signed-off-by: Jean Thomas <jean.thomas@wifirst.fr>

JeThWifirst commented on target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi:
```diff
@@ -0,0 +1,257 @@
+
+&pwm {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pwm0_pin>;
```
This is already enabled in the parent *.dtsi file, see [here](https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/files-5.15/arch/arm64/boot/dts/mediatek/mt7981.dtsi#L44-L54).

JeThWifirst commented on target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi:
```diff
@@ -0,0 +1,257 @@
+
+/ {
+	chosen {
+		bootargs = "console=ttyS0,115200n1 loglevel=8 earlycon=uart8250,mmio32,0x11002000 root=/dev/mmcblk0p7 rootwait rootfstype=squashfs";
```
No, as there is no default command line for this target. But I trimmed the `bootargs` a bit, and now use the `PARTLABEL`

JeThWifirst commented:
> Also - take a look at blocktrron's staging tree - as they are working on this same device if I am right.

That was my fear: someone working on the same device, and both of us doing the work twice. Thanks for the tip, I will check @blocktrron staging tree, I found [the x3000 branch](https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=shortlog;h=refs/heads/x3000).

dangowrt commented on target/linux/mediatek/patches-6.1/251-pinctrl-mediatek-mt7981-fix-emmc_45-group.patch:
```diff
@@ -0,0 +1,34 @@
+ /* EMMC */
+-static int mt7981_emmc_45_pins[] = { 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
+-static int mt7981_emmc_45_funcs[] = { 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, };
++static int mt7981_emmc_45_pins[] = { 0, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
```
So now I took a look at the datasheet and checked what function 2 of GPIO 0 actually does: `WA_AICE_TCKC`. So that's a JTAG clock signal which is intended to be used when debugging the built-in wireless controller. And that smells a bit fishy.
I totally believe that this fixes the issue of the eMMC not working, but we have to find out why. Is GPIO 0 really used as a signal for the eMMC on that board? If so, do we really have to set function 2, or is it enough to drive the pin as output and set it either low or high using a GPIO hog?

JeThWifirst replied:
> So now I took a look at the datasheet and checked what function 2 of GPIO 0 actually does: WA_AICE_TCKC

Sadly I don't have access to the datasheet, and I didn't find it online. Would it be possible to email it to me, if it is not protected by a NDA?

Regarding the eMMC I initially added the GPIO 0 with function 2 because it is what is done in the GL.iNet firmware. However, leaving it unclaimed works on my device.  Maybe the pin is driven by the hardware with a pull-up or a pull-down resistor?

> or is it enough to drive the pin as output and set it either low or high using a GPIO hog?

You mean adding a GPIO hog in the device tree, just in case it is not really HW-driven?
```
&pio {
	gpio_wps-hog {
		gpio-hog;
		gpios = <0 GPIO_ACTIVE_HIGH>;
		output-high;
	};
        [...]
}
```

But there is another issue which will force me to create a new MMC group: the GPIO 15 present in the `emmc_45` group is also used by the `pwm0_1` [group](https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/files-5.15/drivers/pinctrl/mediatek/pinctrl-mt7981.c#L676-L677), which is used to drive the fan with the `pwm0_pin` [pin](https://github.com/JeThWifirst/openwrt/blob/add-gl-inet-xe3000-support/target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi#L195-L200).

So my plan is to rework my second pinctrl patch in order to create a new `mmc` group, without the pins 0 and 15:
```c
static int mt7981_emmc_pins[] = { 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
static int mt7981_emmc_funcs[] = { 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, };
```

Would that solution be okay for you?

JeThWifirst replied:
The last version I pushed implements this solution

dangowrt replied:
> So my plan is to rework my second pinctrl patch in order to create a new `mmc` group, without the pins 0 and 15:

So I've been stairing at the datasheet for a short while and concluded:

* Removing pin 15 can work because that's *EMMC_RSTB*, and maybe the reset of the eMMC just isn't connected or hard-wired to board reset circuit or another pin **in GPIO mode** is can be used to drive the reset signal. That liberates pin 15 of the SoC which can then be used for PWM instead, for example. 

* Setting pin 0 to function 2 doesn't make sense, that would be a the JTAG clock of the wireless MAC co-processor used when debugging the WA firmware using JTAG. If it is really used as reset for the emmc, it should be wired up as `reset-gpios` of the mmc bus in device tree.

* Pin 16~19 will always be used as *EMMC_DAT0*...*EMMC_DAT3*.

* Pin 20~23 are used in case we are using 8-bit bus-width for the eMMC, otherwise they can be liberated as well.

Hence, to me it would make most sense to have pin 15 in a separate group of it's own called emmc_reset, and have emmc4 and emmc8 groups for 4 bit and 8 bit bus-width respectively.

I hope I could give some insights here. Let me know if further assistance is needed so we can make this move forward.

mrkiko commented:
Hi!
@dangowrt have you tested on your MT2500 as well?


On Sun, Dec 10, 2023 at 09:16:41AM -0800, Daniel Golle wrote:
> @dangowrt commented on this pull request.
> 
> 
> 
> > +for the MediaTek MT7981 SoC.
> +
> +Signed-off-by: Jean Thomas ***@***.***>
> +---
> + drivers/pinctrl/mediatek/pinctrl-mt7981.c | 6 +++---
> + 1 file changed, 3 insertions(+), 3 deletions(-)
> +
> +--- a/drivers/pinctrl/mediatek/pinctrl-mt7981.c
> ++++ b/drivers/pinctrl/mediatek/pinctrl-mt7981.c
> +@@ -700,8 +700,8 @@
> + static int mt7981_drv_vbus_funcs[] = { 1, };
> +
> + /* EMMC */
> +-static int mt7981_emmc_45_pins[] = { 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
> +-static int mt7981_emmc_45_funcs[] = { 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, };
> ++static int mt7981_emmc_45_pins[] = { 0, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, };
> 
> So now I took a look at the datasheet and checked what function 2 of GPIO 0 actually does: `WA_AICE_TCKC`. So that's a JTAG clock signal which is intended to be used when debugging the built-in wireless controller. And that smells a bit fishy.
> I totally believe that this fixes the issue of the eMMC not working, but we have to find out why. Is GPIO 0 really used as a signal for the eMMC on that board? If so, do we really have to set function 2, or is it enough to drive the pin as output and set it either low or high using a GPIO hog?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/14142#pullrequestreview-1774068448
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

aiamadeus commented on target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi:
```diff
@@ -0,0 +1,257 @@
+
+/ {
+	chosen {
+		bootargs = "console=ttyS0,115200n1 loglevel=8 earlycon=uart8250,mmio32,0x11002000 root=/dev/mmcblk0p7 rootwait rootfstype=squashfs";
```
`bootargs = "console=ttyS0,115200n8 root=PARTLABEL=rootfs rootwait";` is enough.

JeThWifirst replied:
Indeed, modified

aiamadeus commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -351,6 +351,31 @@ define Device/glinet_gl-mt6000
+define Device/glinet_gl-x3000-xe3000-common
+  DEVICE_VENDOR := GL.iNet
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7981-firmware mt7981-wo-firmware kmod-mmc \
```
This is incorrect, the mmc driver is built-in into the kernel. And `usb-storage` is to support USB Storage devices.

JeThWifirst replied:
Okay, I removed `kmod-mmc`, but for me `kmod-usb-storage` is really necessary to be able to correctly use the SD card or the USB key plugged on the device. I tried to remove it, and without it the USB storage is not correctly detected.

Here are the relevant logs:
```
[    2.658823] usb 1-1.1: new high-speed USB device number 3 using xhci-mtk
[    2.793800] usb-storage 1-1.1:1.0: USB Mass Storage device detected
[    2.800427] scsi host0: usb-storage 1-1.1:1.0
[    3.809999] scsi 0:0:0:0: Direct-Access     Generic  STORAGE DEVICE   0233 PQ: 0 ANSI: 0
[    4.033350] sd 0:0:0:0: [sda] 7761920 512-byte logical blocks: (3.97 GB/3.70 GiB)
[    4.041755] sd 0:0:0:0: [sda] Write Protect is off
[    4.046537] sd 0:0:0:0: [sda] Mode Sense: 0b 00 00 08
[    4.052482] sd 0:0:0:0: [sda] No Caching mode page found
[    4.057782] sd 0:0:0:0: [sda] Assuming drive cache: write through
[    4.068167]  sda: sda1
[    4.073124] sd 0:0:0:0: [sda] Attached SCSI removable disk
[...]
[   23.978823] usb 1-1.3: new high-speed USB device number 5 using xhci-mtk
[   24.111412] usb-storage 1-1.3:1.0: USB Mass Storage device detected
[   24.119273] scsi host1: usb-storage 1-1.3:1.0
[   25.330657] scsi 1:0:0:0: Direct-Access     General  USB Flash Disk   1100 PQ: 0 ANSI: 4
[   25.340250] sd 1:0:0:0: [sdb] 7856128 512-byte logical blocks: (4.02 GB/3.75 GiB)
[   25.348195] sd 1:0:0:0: [sdb] Write Protect is off
[   25.353003] sd 1:0:0:0: [sdb] Mode Sense: 43 00 00 00
[   25.358517] sd 1:0:0:0: [sdb] No Caching mode page found
[   25.363853] sd 1:0:0:0: [sdb] Assuming drive cache: write through
[   25.383373]  sdb: sdb1 sdb2 sdb3
[   25.389071] sd 1:0:0:0: [sdb] Attached SCSI removable disk
```

aiamadeus replied:
`kmod-usb-storage` is required to support usb-storage devices, but is not normally used as a default package.
Device packages are the smallest packages that keep the device running properly.

JeThWifirst replied:
Grepping the code shows quite a few targets having `kmod-usb-storage` listed in `DEVICE_PACKAGES`, which makes sense IMO as it is necessary to offer all the features of the device: a user flashing a vanilla OpenWrt expects to be able to use a SD card or a USB key on the device without having to install an additional kernel module.

But I don't have a strong stance on this, and you have much more experience on OpenWrt repos than me, so if you want me to remove it, I'll remove it.

aiamadeus commented on target/linux/mediatek/dts/mt7981a-glinet-gl-x3000-xe3000-common.dtsi:
```diff
@@ -0,0 +1,241 @@
+		led_power: led@6 {
+			label = "green:power";
+			gpios = <&pio 39 GPIO_ACTIVE_LOW>;
+			default-state = "on";
```
No need to set default-state, this is already controlled by led-xxx (in aliases).
Also you need to replace led@x with function name.

JeThWifirst replied:
`default-state` removed and function names used for LEDs

aiamadeus commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/firmware/11-mt76-caldata:
```diff
@@ -31,7 +31,9 @@ case "$FIRMWARE" in
 		esac
 		;;
-	ubnt,unifi-6-plus)
+	ubnt,unifi-6-plus|\
```
Please keep alphabetical order in all scripts.

JeThWifirst replied:
Right, I did not notice it was alphabetically ordered. I moved a few blocks, it should be okay now

zhaojh329 commented:
@dangowrt 
I noticed that the mhi driver in latest OpenWrt's kernel does not support RM520N-GL.
Should backport it from the latest kernel?

hauke commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -391,6 +391,31 @@ define Device/glinet_gl-mt6000
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7981-firmware mt7981-wo-firmware mkf2fs \
+    kmod-fs-f2fs kmod-hwmon-pwmfan kmod-usb3 kmod-usb-serial-option \
+    kmod-usb-storage kmod-usbip-client kmod-usb-net-qmi-wwan uqmi
```
Why do you need `kmod-usbip-client`?

JeThWifirst replied:
This is for the USB tethering:
```
[    5.823286] vhci_hcd vhci_hcd.0: USB/IP Virtual Host Controller
[    5.829246] vhci_hcd vhci_hcd.0: new USB bus registered, assigned bus number 3
[    5.836505] vhci_hcd: created sysfs vhci_hcd.0
[    5.841389] hub 3-0:1.0: USB hub found
[    5.850412] hub 3-0:1.0: 8 ports detected
[    5.855150] vhci_hcd vhci_hcd.0: USB/IP Virtual Host Controller
[    5.861078] vhci_hcd vhci_hcd.0: new USB bus registered, assigned bus number 4
[    5.868554] usb usb4: We don't know the algorithms for LPM for this host, disabling LPM.
[    5.877202] hub 4-0:1.0: USB hub found
[    5.882761] hub 4-0:1.0: 8 ports detected
```

dangowrt replied:
USB-IP is a protocol to transport USB bus data over TCP/IP, typically used inside LANs with low and constant latency. It's a nice toy to remote access USB printers, scanners and the like for which no Linux drivers exist. Of course, it can also be useful as a development tool. But I fail to understand how it would be a requirement for USB tethering. Or is the modem offering USB-over-IP-over-PCIe when connected via PCIe? (only? is that even possible? I thought USB 2 is always connected...)

JeThWifirst replied:
I may very well have misunderstood the use of `kmod-usbip-client`. I just tried to replicate the USB devices detected by the GL.iNet firmware, and I noticed that without this module, the USB/IP Virtual Host Controllers I pointed out in the logs were not detected. I'll double-check tomorrow on the device, as I don't have it at hand today.

JeThWifirst replied:
I confirm that removing `kmod-usbip-client` removes the detection of these two USB/IP Virtual Host Controllers, but as I'm not sure what they are used for, I prefer dropping this module.

hauke commented:
Did you send the pinctrl patches upstream for inclusion into the mainline linux kernel?

JeThWifirst commented:
> Did you send the pinctrl patches upstream for inclusion into the mainline linux kernel?

No, I did not know this was necessary. But if this is okay with you, I think it would be best to first reach an agreement with @dangowrt on what are the correct pinctrl patches, and then I will submit them to the mainline linux kernel.

dangowrt commented:
Nice, emmc pinctrl groups now make sense. I agree that we should keep existing `emmc_45` group for legacy reasons, but that should probably be mentioned in the commit description because `emmc_8` + `emmc_reset` == `emmc_45`.

JeThWifirst commented:
Thanks a lot for your support and the information from the datasheet!

I created the groups as you suggested, and edited the commit message to explain why `emmc_45` is kept.

dangowrt commented on target/linux/mediatek/patches-5.15/251-pinctrl-mediatek-mt7981-add-additional-emmc-group.patch:
```diff
@@ -0,0 +1,56 @@
+ 	PINCTRL_PIN_GROUP("udi", mt7981_udi),
+ 	/* @GPIO(14) DRV_VBUS(1) */
+ 	PINCTRL_PIN_GROUP("drv_vbus", mt7981_drv_vbus),
++	/* @GPIO(15): EMMC(2) */
```
```suggestion
+	/* @GPIO(15): EMMC_RSTB(2) */
```

JeThWifirst replied:
All suggestions applied

dangowrt commented on target/linux/mediatek/patches-5.15/251-pinctrl-mediatek-mt7981-add-additional-emmc-group.patch:
```diff
@@ -0,0 +1,56 @@
+ 	PINCTRL_PIN_GROUP("drv_vbus", mt7981_drv_vbus),
++	/* @GPIO(15): EMMC(2) */
++	PINCTRL_PIN_GROUP("emmc_reset", mt7981_emmc_reset),
++	/* @GPIO(16,19.24,25): EMMC(2) */
```
```suggestion
+	/* @GPIO(16,17,18,19,24,25): EMMC_DATx, EMMC_CLK, EMMC_CMD */
```

dangowrt commented on target/linux/mediatek/patches-5.15/251-pinctrl-mediatek-mt7981-add-additional-emmc-group.patch:
```diff
@@ -0,0 +1,56 @@
++	PINCTRL_PIN_GROUP("emmc_reset", mt7981_emmc_reset),
++	/* @GPIO(16,19.24,25): EMMC(2) */
++	PINCTRL_PIN_GROUP("emmc_4", mt7981_emmc_4),
++	/* @GPIO(16,25): EMMC(2) */
```
```suggestion
+	/* @GPIO(16,17,18,19,20,21,22,23,24,25): EMMC_DATx, EMMC_CLK, EMMC_CMD */
```

dangowrt commented on target/linux/mediatek/patches-6.1/251-pinctrl-mediatek-mt7981-add-additional-emmc-group.patch:
```diff
@@ -0,0 +1,56 @@
+ 	PINCTRL_PIN_GROUP("udi", mt7981_udi),
+ 	/* @GPIO(14) DRV_VBUS(1) */
+ 	PINCTRL_PIN_GROUP("drv_vbus", mt7981_drv_vbus),
++	/* @GPIO(15): EMMC(2) */
```
```suggestion
+	/* @GPIO(15): EMMC_RSTB(2) */
```

dangowrt commented on target/linux/mediatek/patches-6.1/251-pinctrl-mediatek-mt7981-add-additional-emmc-group.patch:
```diff
@@ -0,0 +1,56 @@
+ 	PINCTRL_PIN_GROUP("drv_vbus", mt7981_drv_vbus),
++	/* @GPIO(15): EMMC(2) */
++	PINCTRL_PIN_GROUP("emmc_reset", mt7981_emmc_reset),
++	/* @GPIO(16,19.24,25): EMMC(2) */
```
```suggestion
+	/* @GPIO(16,17,18,19,24,25): EMMC_DATx, EMMC_CLK, EMMC_CMD */
```

dangowrt commented on target/linux/mediatek/patches-6.1/251-pinctrl-mediatek-mt7981-add-additional-emmc-group.patch:
```diff
@@ -0,0 +1,56 @@
++	PINCTRL_PIN_GROUP("emmc_reset", mt7981_emmc_reset),
++	/* @GPIO(16,19.24,25): EMMC(2) */
++	PINCTRL_PIN_GROUP("emmc_4", mt7981_emmc_4),
++	/* @GPIO(16,25): EMMC(2) */
```
```suggestion
+	/* @GPIO(16,17,18,19,20,21,22,23,24,25): EMMC_DATx, EMMC_CLK, EMMC_CMD */
```

dangowrt reviewed (approved):
If you like squash those tiny changes to the comments describing the pins to be as exact as possible.

JeThWifirst commented:
Thanks a lot for the review @dangowrt!

I have two further questions for you at this stage:
* Should I send the pinctrl patches (UART and eMMC) upstream for mainline linux kernel inclusion, as suggested by @hauke?
* In order for this device to correctly use the modem, I submitted two patches for `uqmi` ([[1]](http://lists.openwrt.org/pipermail/openwrt-devel/2023-December/041900.html) and [[2]](http://lists.openwrt.org/pipermail/openwrt-devel/2023-December/041901.html)) on the OpenWrt developers mailing-list, but I didn't get any reviews. As I see your name in the Git history of this project, could you please have a look at these patches too?

Many thanks in advance

dangowrt commented:
> Should I send the pinctrl patches (UART and eMMC) upstream for mainline linux kernel inclusion, as suggested by @hauke?

Yes, both patches look good and should go to upstream mailing lists please.

> * In order for this device to correctly use the modem, I submitted two patches for `uqmi` ([[1]](http://lists.openwrt.org/pipermail/openwrt-devel/2023-December/041900.html) and [[2]](http://lists.openwrt.org/pipermail/openwrt-devel/2023-December/041901.html)) on the OpenWrt developers mailing-list, but I didn't get any reviews. As I see your name in the Git history of this project, could you please have a look at these patches too?

Merged. Thanks!

JeThWifirst commented:
> Yes, both patches look good and should go to upstream mailing lists please.

I posted the patches to the mailing lists, hopefully correctly as this is the first time I send a patch to the kernel.

> Merged. Thanks!

Thank you for the review, the merge, and the bump of the `uqmi` package!

mrkiko commented:
Hi!!

It would be great if you could indicate how to switch the modem back to PCI / default mode in the commit message.
It would be even better if you could remove instructions completely and see if the modem can be usable in PCI mode after trying to backport the kernel changes.

JeThWifirst commented:
> It would be great if you could indicate how to switch the modem back to PCI / default mode in the commit message. 

I updated the commit message to indicate how to switch back and forth between PCIe and USB interfaces for the modem.

> It would be even better if you could remove instructions completely and see if the modem can be usable in PCI mode after trying to backport the kernel changes.

Based on the work by @blocktrron and his [x3000 branch](https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=shortlog;h=refs/heads/x3000), I tried to use the modem in PCIe mode. The result is available [in this commit](https://github.com/openwrt/openwrt/commit/7dff14e450581693fe6341efc5ee75f1ef53f761): the modem is correctly detected, but I stopped here as I soon realized that everything in OpenWrt is based on a USB-modem: [the wiki](https://openwrt.org/docs/guide-user/network/wan/wwan/ltedongle) describing how to use a LTE modem, [the `qmi.sh` script](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh;hb=HEAD#l71) used to communicate with the modem in QMI mode, [the `mbim.sh` script](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/network/utils/umbim/files/lib/netifd/proto/mbim.sh;hb=HEAD#l70) used to communicate with the modem in MBIM mode, ...

So unless I missed something (which is perfectly possible), using the modem in PCIe mode seems to require a major overhaul of the WWAN part and would be out of the scope of this PR. Therefore I think it would be best to try to merge this PR in its current state, and then if we want to use the modem in PCIe mode, create a dedicated PR. 

Would that be okay with you @mrkiko?

pesintta commented:
> Based on the work by @blocktrron and his [x3000 branch](https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=shortlog;h=refs/heads/x3000), I tried to use the modem in PCIe mode. The result is available [in this commit](https://github.com/openwrt/openwrt/commit/7dff14e450581693fe6341efc5ee75f1ef53f761): the modem is correctly detected, but I stopped here as I soon realized that everything in OpenWrt is based on a USB-modem: [the wiki](https://openwrt.org/docs/guide-user/network/wan/wwan/ltedongle) describing how to use a LTE modem, [the `qmi.sh` script](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh;hb=HEAD#l71) used to communicate with the modem in QMI mode, [the `mbim.sh` script](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/network/utils/umbim/files/lib/netifd/proto/mbim.sh;hb=HEAD#l70) used to communicate with the modem in MBIM mode, ...
> 

Isn't MHI supposed to expose the modem in various modes (QMI, MBIM etc.) under /dev which can be used to communicate with it like with USB device?

I tried your WIP commit of PCIe support and indeed new devices appear:

```
~# ls -l /dev/wwan*
crw-------    1 root     root      250,   1 Jan 21 12:29 /dev/wwan0mbim0
crw-------    1 root     root      250,   0 Jan 21 12:29 /dev/wwan0qcdm0
crw-------    1 root     root      250,   2 Jan 21 12:29 /dev/wwan0qmi0
```

Sadly though I couldn't get uqmi to get a response from `wwan0qmi0`. I don't know if it is because uqmi is doing something wrong or something else is missing from MHI/PCIe bits.

JeThWifirst commented:
> Isn't MHI supposed to expose the modem in various modes (QMI, MBIM etc.) under /dev which can be used to communicate with it like with USB device?

This is also my understanding: we should be able to communicate with the modem using the QMI protocol over one of the new devices exposed by the modem in PCIe mode, as it is possible using `/dev/cdc-wdm0` when the modem is in USB mode.

What I pointed out in my last message is that the OpenWrt scripts try to find the device used to communicate with the modem in `/sys/class/usbmisc/` (so a USB device, not a PCIe one), and that the wiki uses `/dev/cdc-wdm0` in every single page about the modem use.

I'm not saying that the modem cannot be used by OpenWrt over PCIe, I'm just saying that after a quick glance it seems to be quite a big change to introduce PCIe support. And in my opinion, it would be out the initial scope of this PR, which is to introduce support for new GL.iNet devices. That is why I did not invest too much time in trying to make it work using PCIe mode right now.

mrkiko commented:
On Mon, Jan 22, 2024 at 12:54:17AM -0800, JeThWifirst wrote:
> > Isn't MHI supposed to expose the modem in various modes (QMI, MBIM etc.) under /dev which can be used to communicate with it like with USB device?
> 
> This is also my understanding: we should be able to communicate with the modem using the QMI protocol over one of the new devices exposed by the modem in PCIe mode, as it is possible using `/dev/cdc-wdm0` when the modem is in USB mode.
> 
> What I pointed out in my last message is that the OpenWrt scripts try to find the device used to communicate with the modem in `/sys/class/usbmisc/` (so a USB device, not a PCIe one), and that the wiki uses `/dev/cdc-wdm0` in every single page about the modem use.
> 
> I'm not saying that the modem cannot be used by OpenWrt over PCIe, I'm just saying that after a quick glance it seems to be quite a big change to introduce PCIe support. And in my opinion, it would be out the initial scope of this PR, which is to introduce support for new GL.iNet devices. That is why I did not invest too much time in trying to make it work using PCIe mode right now.

In my opinion, since the commit mentions how to put the modem back in it's normal state, this PR is ready to be merged in absence of other issues.
Merging the code will also increase (or - at least not decrease) the chances of more interested people looking at this and trying to implement it.
Being the performance factor considered really important both here and in gl.iNet forums, maybe we can note in the commit the PCI mode should perform better even tough not yet supported, just to be clear and to avoid comparing apples to oranges when it comes to performance data.

Just my 2 cents.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/14142#issuecomment-1903522597
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

JeThWifirst commented:
Hi everyone,

Good news, the pinctrl patches have been merged upstream:
* https://git.kernel.org/pub/scm/linux/kernel/git/linusw/linux-pinctrl.git/commit/?h=for-next&id=398f914bf0022dcde82050aab72a2bd842ee1709
* https://git.kernel.org/pub/scm/linux/kernel/git/linusw/linux-pinctrl.git/commit/?h=for-next&id=8714b3414dd6070609fabdd73a70e5f818f77f24

> In my opinion, since the commit mentions how to put the modem back in it's normal state, this PR is ready to be merged in absence of other issues. Merging the code will also increase (or - at least not decrease) the chances of more interested people looking at this and trying to implement it.

This is also my opinion: this PR has now been open for almost two months, and as far as I can see there are no blocking points preventing it to be merged. Several reviewers have checked the code (many thanks to you!), and [the last open point](https://github.com/openwrt/openwrt/pull/14142/#discussion_r1426680919) is a small detail, but I'm still waiting for an answer from @aiamadeus. Merging the code would allow to have two new functional devices supported by OpenWrt, which can still be improved in the future when more people use them and report bugs, or contribute to improve them. 

Please let me know if I can be of any assistance to finalize this PR before merging it!

JeThWifirst commented:
I rebased the PR on top of the latest master in order to make it easier to review.

I also mirrored the changes introduced by @dangowrt for GL.iNet MT6000 (bf205854128cc791e5805ed20ef3bd0bb53c339a) in order to get rid of Wi-Fi EEPROM and MAC address setup in userspace, which is much cleaner, thanks for that!

mrkiko commented:
On Mon, Feb 19, 2024 at 05:32:50AM -0800, JeThWifirst wrote:
> I rebased the PR on top of the latest master in order to make it easier to review.
> 
> I also mirrored the changes introduced by @dangowrt for GL.iNet MT6000 (bf205854128cc791e5805ed20ef3bd0bb53c339a) in order to get rid of Wi-Fi EEPROM and MAC address setup in userspace, which is much cleaner, thanks for that!
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/14142#issuecomment-1952463385
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

Hi!
Thanks for keeping this going! :)

BTW - are you experiencing as well the overlay being extended to use the whole space of the eMMC, differently than what happened before?
Trying to understa why it happens but no clear idea for now.

JeThWifirst commented:
> BTW - are you experiencing as well the overlay being extended to use the whole space of the eMMC, differently than what happened before? Trying to understa why it happens but no clear idea for now.

Not sure what you are referring to, could you please elaborate a bit?

mrkiko commented:
On Mon, Feb 19, 2024 at 09:04:17AM -0800, JeThWifirst wrote:
> > BTW - are you experiencing as well the overlay being extended to use the whole space of the eMMC, differently than what happened before? Trying to understa why it happens but no clear idea for now.
> 
> Not sure what you are referring to, could you please elaborate a bit?

Sorry if I've not been clear.
Before, i could experiment OpenWrt using a portion of the eMMC as rootfs + overlay, while now it uses the full 8GB of the eMMC on my gl-mt2500. Wondering if you observed that difference too and maybe you had an idea of why this happens.

Thanks.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/14142#issuecomment-1952890643
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

JeThWifirst commented:
> Before, i could experiment OpenWrt using a portion of the eMMC as rootfs + overlay, while now it uses the full 8GB of the eMMC on my gl-mt2500. Wondering if you observed that difference too and maybe you had an idea of why this happens.

No, I did not noticed such a change on GL.iNet X3000, no idea why it happens on your device

mocmocamoc commented:
Would be great to get this merged if the reviewers could take a look at it please.

mocmocamoc commented:
@dangowrt @hauke @aiamadeus This PR seems ready to go but has been stuck for a while, please could you take a look at merging it if no objections?

dangowrt commented on target/linux/mediatek/dts/mt7981a-glinet-gl-xe3000.dts:
```diff
@@ -0,0 +1,8 @@
+
+/ {
+	model = "GL.iNet GL-XE3000";
+	compatible = "glinet,gl-xe3000", "mediatek,mt7981";
```
So the only difference is the name?!

In the case, please use a single image and add `DEVICE_ALT_*` fields in `target/linux/mediatek/image/filogic.mk`.

JeThWifirst replied:
There are no differences in the device tree between the two devices, except for the name, yes. However, the XE3000 device has a battery, while the X3000 does not have one. 

As far as I can see, merging both devices in a single image using `DEVICE_ALTX_*` fields would mean having the same identifier in `/proc/device-tree/compatible` and `/proc/device-tree/model`: wouldn't it be an issue for the users trying to identify the devices after flashing the image?

But I'm not familiar with the way such similar devices are handled in OpenWrt, so if you confirm that I should use a single image for both devices, I will do the change.

dangowrt replied:
The general rule is: If there are no software differences justifying the need for another image (and hence continuous need to spend resources on build and download servers), then we shouldn't create another image.
If there is a way to interact with the PMIC/charger e.g. via I2C on the XE3000 device (ie. to know the current battery level, charging/discharging current, ...), then of course that I2C device should be defined in DT and would be such a software difference.

dangowrt replied:
Ack. I don't see any USB involved there, but never mind. In the long run it would be better to implement this as a serdev driver in kernel, exposing the usual hwmon interface instead of having a custom tool in userspace, but for now this is fine.

mocmocamoc replied:
Please could you mark the request change as addressed so that this can be merged?

dangowrt commented on target/linux/mediatek/patches-6.1/251-pinctrl-mediatek-mt7981-add-additional-emmc-group.patch:
```diff
@@ -0,0 +1,56 @@
+mediatek: mt7981: pinctrl: add additional emmc groups
```
Has this been sent upstream?
If so, please add `X-Patchwork-*` headers (along with other RFC822 headers which would be needed to re-send the patch upstream)

JeThWifirst replied:
Yes, the patch has been merged upstream in linusw tree (https://git.kernel.org/pub/scm/linux/kernel/git/linusw/linux-pinctrl.git/commit/?h=for-next&id=398f914bf0022dcde82050aab72a2bd842ee1709). I've taken the patch from https://patchwork.kernel.org in order to add the requested headers, please let me know if any are missing.

dangowrt commented on target/linux/mediatek/patches-6.1/250-pinctrl-mediatek-mt7981-add-additional-uart-group.patch:
```diff
@@ -0,0 +1,40 @@
+mediatek: mt7981: pinctrl: add additional uart group
```
Has this been sent upstream?
If so, please add `X-Patchwork-*` headers (along with other RFC822 headers which would be needed to re-send the patch upstream)

JeThWifirst replied:
Yes, the patch has been merged upstream in linusw tree (https://git.kernel.org/pub/scm/linux/kernel/git/linusw/linux-pinctrl.git/commit/?h=for-next&id=8714b3414dd6070609fabdd73a70e5f818f77f24). I've taken the patch from https://patchwork.kernel.org in order to add the requested headers, please let me know if any are missing.

JeThWifirst commented:
@dangowrt I also renamed the patches in order to avoid a number collision with existing patches, and added the patches for the 6.6 kernel, as the upstream changes are not yet integrated.

Kilianos709 commented:
Thanks to all you folks for all the hard work on this. We will all be happy to see this device fully supported. It will be great to have full control and transparency!

Hootan commented:
Hello, do you know if there is any way to remove Russia/Iran ban from this firmware? 🙏
